### ISSUE TITLE
VolPy Mask_RCNN doesn't work

### ISSUE THREAD

**Issue by bobqianic** [flatironinstitute/CaImAn#1567]
My supervisor gave me a piece of code. I ran it on the latest version of CaImAn, but Mask R-CNN consistently produces no ROIs, while it works well with the TensorFlow version `utils.mrcnn_inference` using the same config.

 ```python
    ############ Segmentation

    # Create mean and correlation images
    img = mean_image(mc.mmap_file[0], window = 1000, dview=dview)
    img = (img-np.mean(img))/np.std(img)


    Cn = local_correlations_movie_offline(mc.mmap_file[0], fr=fr, window=fr*4, #### [0] ???
                                          stride=fr*4, winSize_baseline=fr, 
                                          remove_baseline=True, gaussian_blur=gaussian_blur,
                                          dview=dview).max(axis=0)
    img_corr = (Cn-np.mean(Cn))/np.std(Cn)
    summary_images = np.stack([img, img, img_corr], axis=0).astype(np.float32)

    # Save summary images which could be further used in the VolPy GUI
    cm.movie(summary_images).save(cells_savedir+ '//summary_images.tif')
    print(cells_savedir+ '//summary_images.tif')

    use_maskrcnn = True  # set to True to predict the ROIs using the mask R-CNN
    if not use_maskrcnn:                 # use manual annotations
        with h5py.File(path_ROIs, 'r') as fl:
            ROIs = fl['mov'][()]  # load ROIs
    else:
        weights_path = "/root/caiman_data/model/mask_rcnn"  
    
        ROIs = utils.mrcnn_inference_pytorch(img=summary_images.transpose([1, 2, 0]), size_range=size_range,#[5, 22], size_range=[5,45]
                                      weights_path=weights_path, display_result=True) # size parameter decides size range of masks to be selected
        cm.movie(ROIs).save(cells_savedir + '//mrcnn_ROIs.hdf5')
        print(cells_savedir + '//mrcnn_ROIs.hdf5')
```

**Comment by bobqianic** [comment]
The summary images produced are identical

```
sha256sum

26cb7dd32991d4b734d1744bfb132227e88632f63a1829a09af56ae7c5ab80c0  latest
26cb7dd32991d4b734d1744bfb132227e88632f63a1829a09af56ae7c5ab80c0  old
```

**Comment by pgunn** [comment]
@manuelpaeza Can you give this a look?

**Comment by bobqianic** [comment]
`demo_pipeline_voltage_imaging.ipynb`

PyTorch-CalmAn 1.13.1:

<img width="845" height="432" alt="Image" src="https://github.com/user-attachments/assets/69195923-d6cc-4c3d-8438-aae9c9bb74e7" />

Tensorflow-CalmAn 1.12.2:

<img width="1252" height="1272" alt="Image" src="https://github.com/user-attachments/assets/0581cae4-5cf2-4919-b649-88f46f3720b5" />

**Comment by bobqianic** [comment]
@pgunn Thank you for looking into this issue. This problem cost me two days and caused quite a bit of confusion. Initially, I tested the demo and saw that there was some ROIs, but it couldn’t identify all cells. I thought it was a threshold issue. However, when I tried it on my experimental data, there was zero ROI.

I suspected there might be a bug in my code, so I asked my supervisor for his code, but it still didn’t work. That’s when I realized there must be a bug in the package itself.

I didn’t expect this to happen in this project. I thought this was a well-established and actively maintained piece of software with over 700 stars. I’ll look into this further, try to identify the problem, and propose a pull request. I’ll also add this to the CI to prevent similar regressions in the future.

**Comment by pgunn** [comment]
@bobqianic Ah, sorry about that; the rewrite of the machine learning code from tensorflow to pytorch touched a lot of components of the codebase, and volpy is unfortunately not the most heavily-maintained or well-understood part of the codebase (it was dropped off by a contributor some years back and is not really fully integrated into the rest of caiman). We do our best, but volpy is a weak spot in terms of maintenance and support.

I hope longer-term we can find a way to integrate it better with the rest of the code and improve its testing.

**Comment by pgunn** [comment]
@bobqianic The person who ported the code to torch has a possible fix which he's testing for comparison; if things look good we may be able to merge a fix into dev soon.

**Comment by pgunn** [comment]
@bobqianic Could you give this branch a try with your data?
https://github.com/flatironinstitute/CaImAn/pull/1568

**Comment by bobqianic** [comment]
@pgunn I just tested that branch on demo data. As you can see, the result still looks off, there’s more ROI, but some are in the wrong position. This didn’t happen in the TF version (see image I posted earlier)

<img width="859" height="594" alt="Image" src="https://github.com/user-attachments/assets/44939037-1c57-4b3b-bd9d-14ff49fe9cf6" />

**Comment by bobqianic** [comment]
I also tested this on my experimental data. It looks like we’re only getting one ROI? The mask shape is (400, 512), but it’s hard to tell because there are some internal errors (shape mismatch) in the new code. Either way, it’s still way off compared to the TF version, I can get 16 correct ROIs there.

<img width="1840" height="761" alt="Image" src="https://github.com/user-attachments/assets/ed02197e-67e7-416b-8587-ac90a1a21583" />

**Comment by manuelpaeza** [comment]
Hello. Apologies for the lateness on this, I will be looking at this and giving you an update soon. If you have anything additional to share, please let me know.

**Comment by bobqianic** [comment]
> Hello. Apologies for the lateness on this, I will be looking at this and giving you an update soon. If you have anything additional to share, please let me know.

I recently tried to replicate the Mask R-CNN model used in the VolPy paper using the provided dataset:
https://zenodo.org/records/4515768

However, I was unable to reproduce the reported results in paper. 
https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1008806&rev=2

I followed all the instructions in the paper, including the optimizer settings, learning rate, 128x128 cropping, and data augmentation.

I used PyTorch’s built-in `maskrcnn_resnet50_fpn` as the starting point, training the heads (fpn, rpn, box, mask) for 20 epochs, and then training layer3, layer4, and the heads (fpn, rpn, box, mask) for another 20 epochs. Despite this, the results were poor, the resulting F1 score is quite low, especially on the L1 dataset, HPC>TEG>L1, which is contrary to what is reported in the paper.

Since the original TensorFlow weights perform well, it is unlikely that the results in the paper are fabricated. I suspect there may be some missing details or implementation differences, but I haven’t been able to identify them yet. I will update this post if I find anything useful.

**Comment by bobqianic** [comment]
@caichangjia Hi Changjia, do you have any idea why this happens? In your paper, you mention that only 29 neurons from the L1 dataset are needed to achieve `0.876` on the validation dataset. How is that possible?

<img width="810" height="385" alt="Image" src="https://github.com/user-attachments/assets/2b98d565-0ccf-4264-bf6f-180b6792b421" />

<img width="3000" height="2430" alt="Image" src="https://github.com/user-attachments/assets/63d7b3b8-36ee-4f56-803a-cbde6b6291f1" />

**Comment by bobqianic** [comment]
@pgunn @manuelpaeza 
I just realized that the segmentation doesn’t have to rely on a custom-trained Mask R-CNN. Even though we generate artificial summary images using mean and correlation, these images still contain common cellular features. Some frontier models have channel-invariant properties, so they don’t care about the order of the channels. So I tried using `Cellpose-SAM`, and the results are surprisingly good.

Even without fine-tuning, we can already identify most ROIs:

<img width="1496" height="721" alt="Image" src="https://github.com/user-attachments/assets/a878cd2f-4c62-4d1d-97fc-0694bb096c17" />

So, in the worst case. Even if the main author @caichangjia doesn’t respond, we can simply replace the segmentation with `Cellpose-SAM` and add a note explaining that VolPy’s Mask R-CNN is no longer working. We can mention that it has been replaced with `Cellpose-SAM`, and that if users want to use the original VolPy Mask R-CNN, they should use an older TensorFlow version.
